### ISSUE TITLE
Hubot Brain set/get Does Not Persist Between Scripts

### ISSUE THREAD

**Issue by richardwlu** [RocketChat/hubot-rocketchat#112]
I have a script that sets a userId `robot.brain.set('userId', body.data.userId)` to the brain on start up but when I try to access that value from a different script `robot.brain.get('userId')` I receive a `null`. I can only get the value within the same script.

Is this the intended behavior? I'm looking for a way to set a key/value of userId and auth token after making an http call to the Rocket.Chat API so I can make calls to list room and send messages using the token between different scripts so I only have to login once hitting `/api/login`

**Comment by geekgonecrazy** [comment]
@richardwlu we don't really control anything with the hubot brain.  That's all hubot specific.  I believe to persist though you need to hook up something like redis for the brain to store in.

**Comment by lucasvanhalst** [comment]
You can use the MongoDB database that's already used for Rocket.Chat for this, using this plugin: https://www.npmjs.com/package/hubot-mongodb-brain
Just add `hubot-mongodb-brain` to your `EXTERNAL_SCRIPTS` list in the config and you should be good.

**Comment by geekgonecrazy** [comment]
@lucasvanhalst that sounds like a great solution!  Thanks for sharing!

**Comment by timkinnane** [comment]
I'm trying to get the approach suggested by @lucasvanhalst working. Currently getting connection errors in development using mongo localhost. I'm guessing because the suggested script tries to start its own instance of Mongo, but the ports are in use already by Rocket.chat or it needs further config to connect to the Rocket.chat mongo. If anyone has got this working, please post solutions here, because the hubot-mongodb-brain package is not Rocket.chat specific and is pretty light on docs.

**Comment by lucasvanhalst** [comment]
@timkinnane 
using docker-compose your docker-compose.yml would look look this
```
mongo:
  image: mongo
  volumes:
    - ./data/runtime/db:/data/db
    - ./data/dump:/dump
  command: mongod --smallfiles --oplogSize 128
hubot:
  image: rocketchat/hubot-rocketchat:latest
  environment:
    - ROCKETCHAT_URL=https://hostname.com
    - ROCKETCHAT_USER=bot
    - ROCKETCHAT_PASSWORD=<password>
    - LISTEN_ON_ALL_PUBLIC=true
    - RESPOND_TO_DM=true
    - BOT_NAME=bot
    - MONGODB_URL=mongodb://mongo/hubot-brain
    - EXTERNAL_SCRIPTS=hubot-mongodb-brain,hubot-help,hubot-seen,hubot-diagnostics

```

**Comment by timkinnane** [comment]
Thanks @lucasvanhalst, I actually needed it to work for local development (not in docker) which is why the mongo settings were failing me.

Figured out for local dev, to use the Rocket.chat mongo server I needed:
`export MONGODB_URL=mongodb://127.0.0.1:3001/hubot-brain`

Note also for anyone else doing local dev on multiple hubots, change `hubot-brain` to something unique for each bot or they'll all use the same brain (which is probably bad).

**Comment by timkinnane** [comment]
Have tried using [hubot-mongodb-brain](https://github.com/shokai/hubot-mongodb-brain) and found pros and cons.

TL;DR It's not as cut and dry as it looks, but still good.

### CONS

For local dev, each bot needs to be pointed to the Rocket.Chat mongo.
`export MONGODB_URL=127.0.0.1:3001/hubot-brain`. "hubot-brain" sets the DB name, should be changed per bot if multiple.
If not set, it will try start a new mongo host and conflict with the Rocket.Chat mongo.

For security, Hubot brain data stores should have their own container, not having direct access to Rocket.Chat data. Imagine a poorly coded script deletes a user for some behaviour appropriate to the script, but then the user is gone from all of Rocket.Chat. Nightmare.

For consistency, any connectors between Hubot brains and Rocket.Chat data should come from Meteor package helpers added to Rocket.Chat or the adaptor.

Once deployed (Docker, Heroku etc), Hubot instances are isolated from Rocket.Chat, so they run their own Mongo (no gain in integration), unless configured to connect to the external host Rocket.chat Mongo (no gain in simplicity).

For those reasons, there's no real gain in simplicity, integration or security between Mongo and Redis.

### PROS

Using Mongo for both Rocket.Chat and Hubot means one less thing in the stack.

If using mLab in Heroku for instance, both DBs can be mapped to the same account / data store (e.g. AWS).

Any third party integrations / admin consoles etc could access both data stores with the same approach.

For those reasons, we're replacing all Hubot Redis brains with Mongo.

**Comment by bios-ben** [comment]
@timkinnane what do you mean the bot needs to be pointed to the Rocket.Chat mongo?
I have it setup that Rocket.Chat is on 3000, Hubot is on 3001, and Mongo is on 27017. I get no errors in the log and a nice "MongDB connected" message. If I change the export MONGODB_URL to be 3001 or 3000 instead of 27017 it fails to boot. Just to confirm I'm using the following the hubot script:
export MONGODB_URL=mongodb://localhost:27017/hubot-brain

Regardless, after setting this all up it still doesn't persist. I'm not using docker, just running a native install of everything so can't just configure like @lucasvanhalst

**Comment by timkinnane** [comment]
@bios-ben Perhaps the difference is that I'm not using Docker, are you? That would account for the different ports. I noted above, this is my setup for local development. If you're using Docker you might need some further config to make the data store persistent when the container restores after shut down.

**Comment by bios-ben** [comment]
@timkinnane that's why I was asking you, I have manually setup everything for production without a docker environment ... I figured if it was pointed to mongo that's all it'd need, the Rocket.Chat itself works correctly and saves after reload fine, just the bot brain doesn't at all, so I was wondering if there's more than just pointing it to the port mongo is listening on (default 27017).

**Comment by timkinnane** [comment]
Sorry then @bios-ben, afraid I don't know enough to help you. I've used robomongo to connect to the mongo and visually debug the brain data in operation, that might help? Perhaps also try setting the environment level `HUBOT_LOG_LEVEL=debug` to make sure Hubot isn't having connection problems.

**Comment by davidleong-lithium** [comment]
Having the same problem as @bios-ben.  I have rocketchat (port 3000) and mongo (bound to localhost and docker0, port 27017) and hubot in a docker container.  I get "INFO MongoDB connected" but nothing is ever saved (eg. my pagerduty email for hubot-pagerduty)

**Comment by timkinnane** [comment]
Closing this as it relates to core Hubot operations, not the adapter.
